ratpack-pac4j/src/test/groovy/ratpack/pac4j/internal/Pac4jAuthenticatorSpec.groovy

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ratpack.pac4j.internal
+
+import ratpack.handling.Context
+import ratpack.pac4j.RatpackPac4j
+import ratpack.path.PathBinding
+import ratpack.server.PublicAddress
+import ratpack.server.internal.ConstantPublicAddress
+import spock.lang.Specification
+import spock.lang.Unroll
+
+import static ratpack.pac4j.RatpackPac4j.DEFAULT_AUTHENTICATOR_PATH
+
+class Pac4jAuthenticatorSpec extends Specification {
```

Comment:
Yeah I deliberated about whether it was useful or not and I know you've said before that you don't like or agree with mocking. However :) I found OpenIdRpSpec was not expressing the cases I wanted to test for clearly - I had hoped to make it easier to understand what was happening with callback url generation (tbh it really should be building a URL rather than a String - but that was a wider change than I wanted to cause). Regarding the mocking, this is stubbing without order or interaction verification - I just wanted to get a PublicAddress and PathBinding in there to exercise the different variants. I realize there is still maintenance downside if the public contract for Context.get(Class), PathBinding.getBoundTo() and RatpackPac4j.ClientsProvider.get(Context) changes, but I figured I'd put it up there for your consideration.